warehouse/packaging/metadata_verification.py

```diff
@@ -41,6 +56,109 @@ def _verify_url_pypi(url: str, project_name: str, project_normalized_name: str)
     )
 
 
+def _get_url_content(
+    resolved_ip: str, url: rfc3986.api.URIReference, max_length_bytes: int
+) -> bytes:
+    """
+    Get at least `max_length_bytes` of the contents of a URL
+
+    Raises `GetContentError` on any errors while trying to
+    access the URL.
+    """
+    try:
+        http = urllib3.HTTPSConnectionPool(
+            host=resolved_ip,
+            port=443,
+            headers={
+                "Host": url.host,
+                "User-Agent": "warehouse/1.0 (PyPI URL verifier, https://pypi.org)",
+            },
+            server_hostname=url.host,
+            assert_hostname=url.host,
+            retries=False,
+            timeout=5,
+        )
+        r = http.request("GET", url.path if url.path else "/", preload_content=False)
+    except Exception:
+        raise GetContentError()
+
+    content = r.read(max_length_bytes)
+
+    # When using `preload_content=False`, the HTTP connection must be manually closed
+    r.close()
+
+    return content
+
+
+def _verify_url_meta_tag(
+    url: str, project_name: str, project_normalized_name: str
+) -> bool:
+    user_uri = rfc3986.api.uri_reference(url).normalize()
+
+    # Require the presence of "https" and the host name. Don't allow
+    # the presence of ports
+    if user_uri.port is not None:
+        return False
+    validator = (
+        rfc3986_validators.Validator()
+        .require_presence_of("scheme", "host")
+        .allow_schemes("https")
+    )
+
+    try:
+        validator.validate(user_uri)
+    except rfc3986ValidationError:
+        return False
+
+    # Only allow registered names for the host (e.g.: google.com, api.github.com),
+    # don't allow IP addresses.
+    if IPv4_MATCHER.fullmatch(user_uri.host) or IPv6_MATCHER.fullmatch(user_uri.host):
+        return False
+
+    # The domain name should not resolve to a private or shared IP address
+    try:
+        address_tuples = socket.getaddrinfo(user_uri.host, user_uri.port)
+        if len(address_tuples) == 0:
+            return False
+        for family, _, _, _, sockaddr in address_tuples:
+            ip_address: ipaddress.IPv4Address | ipaddress.IPv6Address | None = None
+            if family == socket.AF_INET:
+                ip_address = ipaddress.IPv4Address(sockaddr[0])
+            elif family == socket.AF_INET6:
+                ip_address = ipaddress.IPv6Address(sockaddr[0])
+            if ip_address is None or not ip_address.is_global:
+                return False
+    except (socket.gaierror, ipaddress.AddressValueError):
+        return False
+
+    # Use the first IP address returned by `getaddrinfo`
+    _, _, _, _, resolved_sockaddr = address_tuples[0]
```

Comment:
nit: You can unpack all in one here: `_, _, _, _, (resolved_ip, _) = address_tuples[0]`